### DISCUSSION TITLE
Stale Closures

### DISCUSSION THREAD

**Discussion by alexchudnovsky** [pmndrs/zustand#784] [Q&A] [answered]
Hi Guys,

Is there any way to get the latest state from a closure ? 

With useState for example you can use setMethod((PreviousValue)=> return PreviousValue + 1)

Is there something like this that can be done with zustand to get the very latest state value without getting the state closure values ?

**Comment by dai-shi** [comment, marked as answer]
You can do:
```js
  set(previousState => ...)
```
Also, you can do `useStore.getState()`.

I think I can give some more examples, if you provide a minimal reproduction.

**Reply by alexchudnovsky** [reply]
ahh thanks

I am declaring this at the top of the react native function 

const { user, loadUserFromAsyncStorage} = userState();

Then inside the useEffect method, I have some inner functions - which are listeners, that are using a value of user

I can not set user as a dependancy on the useEffect as too many things happen inside the useEffect I can only call it once 

But I am not sure how to access the user option for them listener, as I am getting stale value 

with useState I can use setUser((PreviousValue)=> { I have access to the previous value })

But I don't know how to do the equivalent in Zustand

Many thanks for your help

**Reply by dai-shi** [reply]
Can you show how your zustand store looks like?

**Reply by dai-shi** [reply]
You can reply in a thread like this.

Anyway, what you want is: `const currentState = userState.getState()`.

(ah, maybe replying with email doesn't work for threads.)

**Reply by dai-shi** [reply]
btw, it's not equivalent to `React.useState`. React state can't do it.

**Reply by alexchudnovsky** [reply]
Sorry bit new to this.. just deleted my posts below and now replying in the thread :) 

Just wanted to say thanks again... I have used it in my code, solved the problem perfectly :)

**Reply by dai-shi** [reply]
No worries. Good to hear that it's solved.

**Comment by alexchudnovsky** [comment]
sure

export const userState = create((set, get) => ({

    user: null,
    deviceToken: "", //temporary will assign a real one here
    pushMessageJobId: null,

    setDeviceToken: (input) => set({ deviceToken: input }),
    setPushMessageJobId: (input) => set({ pushMessageJobId: input }),



    loadUserFromAsyncStorage: async () => {

        if (!get().user)

            try {
                let storedUser = JSON.parse(await AsyncStorage.getItem(API.asyncStorageKey))

                console.log("The stored user is", storedUser)

                //we don't care about the data which is returned, we just want to ensure that the jwt token is still valid

                await axios.get(`${API.server}/users/me`, {

                    headers: {
                        Authorization:
                            `Bearer ${storedUser.jwt}`,
                    }

                })

                set({ user: storedUser })

                return "complete"

            } catch (e) {
                console.log("Could not load user data from active storage either because he does not exist or the jwt token is invalid", e.message)
            }
    },

**Comment by alexchudnovsky** [comment]
const App = () => {

    const { user, loadUserFromAsyncStorage, setDeviceToken, setPushMessageJobId } = userState();
    let [isLoading, setLoading] = useState(true)

    const navigatorRef = React.useRef()

    useEffect(() => {

        loadUserAndPrepareMessaging()

        setTimeout(() => {
            SplashScreen.hide()
        }, 300)

    }, [])


    async function loadUserAndPrepareMessaging() {

        try {

            await loadUserFromAsyncStorage()
            console.log("the user is displaying null", user)
            await requestUserPermission()

        } catch (error) {

            console.log("Error loading user from storage", error.message)

        } finally {
            setLoading(false)
        }

    }
}

**Reply by alexchudnovsky** [reply]
sorry i can't get the formatting right.. but I hope this gives you an idea 

Thanks again

**Reply by alexchudnovsky** [reply]
The issue is with console.log("the user is displaying null", user)

Even though, it's actually set.. i think its just a stale closure value
